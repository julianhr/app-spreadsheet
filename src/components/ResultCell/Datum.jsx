/* @jsx jsx */
import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'
import { jsx, css } from '@emotion/core' // eslint-disable-line


const Root = styled.div`
  display: flex;
  align-items: center;
  border: 2px solid transparent;
  line-height: 1.1em;
  height: 100%;
  width: 100%;
  padding: 2px;
`

function Datum({
  fwdRef,
  isActive,
  location,
  onClick,
  onDoubleClick,
  onFocus,
  onKeyDown,
  result,
}) {

  const getStyle = () => {
    let style = {}

    if (typeof result === 'number') {
      style = {
        whiteSpace: 'nowrap',
        textOverflow: 'clip',
        overflow: 'hidden',
      }
    }

    if (isActive) {
      style.border = '2px solid salmon'
    }

    return style
  }

  return (
    <Root
      ref={fwdRef}
      data-cell='result'
      data-location={location}
      css={getStyle()}
      tabIndex='0'
      onClick={onClick}
      onKeyDown={onKeyDown}
      onDoubleClick={onDoubleClick}
      onFocus={onFocus}
    >
    {result}
  </Root>
  )
}

Datum.propTypes = {
  fwdRef: PropTypes.object.isRequired,
  result: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]).isRequired,
  isActive: PropTypes.bool.isRequired,
  onClick: PropTypes.func.isRequired,
  onKeyDown: PropTypes.func.isRequired,
  onDoubleClick: PropTypes.func.isRequired,
  onFocus: PropTypes.func.isRequired,
}

export default Datum
