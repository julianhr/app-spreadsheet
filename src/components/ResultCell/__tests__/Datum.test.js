import React from 'react'
import { render, fireEvent, cleanup } from '@testing-library/react'
import emotionSerializer from 'jest-emotion'

import MockApp from '~/__tests__/__mocks__/MockApp'
import Datum from '../Datum'


const requiredProps = {
  fwdRef: React.createRef(),
  result: 5,
  onClick: jest.fn(),
  onKeyDown: jest.fn(),
  onDoubleClick: jest.fn(),
  onFocus: jest.fn(),
  isActive: true,
}

const testProps = {
  ...requiredProps,
}

const renderApp = (props) => {
  const wrapper = render(
    <MockApp>
      <Datum {...props} />
    </MockApp>
  )

  return wrapper
}

describe('Datum', () => {
  afterEach(cleanup)

  test('required props', () => {
    Object.keys(requiredProps).forEach(key => {
      const props = {...testProps}
      delete props[key]
      expect(() => renderApp(props)).toThrow()
    })
  })

  test('result is string', () => {
    const props = { ...testProps }
    props.result = 'test'
    renderApp(props)
    expect.addSnapshotSerializer(emotionSerializer)
    expect(document.querySelector('[data-cell="result"]')).toMatchSnapshot()
  })

  test('result is number', () => {
    const props = { ...testProps }
    props.result = 5
    renderApp(props)
    expect.addSnapshotSerializer(emotionSerializer)
    expect(document.querySelector('[data-cell="result"]')).toMatchSnapshot()
  })
})
