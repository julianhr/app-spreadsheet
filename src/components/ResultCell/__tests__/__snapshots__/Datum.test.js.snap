// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Datum result is number 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 2px solid transparent;
  line-height: 1.1em;
  height: 100%;
  width: 100%;
  padding: 2px;
  white-space: nowrap;
  text-overflow: clip;
  overflow: hidden;
  border: 2px solid salmon;
}

<div
  class="emotion-0 emotion-1"
  data-cell="result"
  tabindex="0"
>
  5
</div>
`;

exports[`Datum result is string 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 2px solid transparent;
  line-height: 1.1em;
  height: 100%;
  width: 100%;
  padding: 2px;
  border: 2px solid salmon;
}

<div
  class="emotion-0 emotion-1"
  data-cell="result"
  tabindex="0"
>
  test
</div>
`;
